src/io.jl

```diff
@@ -223,9 +226,545 @@ function readg(gfile; set_time=nothing)
+    gs::Vector{GEQDSKFile},
+    eq::IMASdd.equilibrium;
+    wall=nothing,
+    geqdsk_cocos::Int=0,
+    dd_cocos::Int=11,
+    cocos_clockwise_phi::Bool=false,
+    add_derived::Bool=false,
+)
+    dd = IMASdd.top_dd(eq)
+    if ismissing(dd.dataset_description.data_entry, :pulse)
+        dd.dataset_description.data_entry.pulse = parse(Int, split(split(gs[1].file, ".")[end-1], "g")[2])
+    end
+    if wall !== nothing
+        geqdsk2wall!(gs[1], wall; geqdsk_cocos=geqdsk_cocos, dd_cocos=dd_cocos, cocos_clockwise_phi=cocos_clockwise_phi)
+    end
+    nt = length(gs)
+    if length(eq.time_slice) == 0  # Blank is okay, but if populated, it had better be compatible
+        resize!(eq.time_slice, nt)
+        eq.time = [g.time for g in gs]
+        for i in 1:nt
+            eq.time_slice[i].time = eq.time[i]
+        end
+    end
+
+    # Write top level data
+    eq.time = [g.time for g in gs]
+    bb = zeros(length(gs))
+    for i in 1:length(gs)
+        if geqdsk_cocos == 0
+            g = gs[i]
+            geqdsk_cocos_ = CoordinateConventions.identify_cocos(
+                sign(g.bcentr), sign(g.current), sign(g.qpsi[1]), sign(g.psi[end] - g.psi[1]), cocos_clockwise_phi,
+            )[1]
+        else
+            geqdsk_cocos_ = geqdsk_cocos
+        end
+        tc = CoordinateConventions.transform_cocos(geqdsk_cocos_, dd_cocos)
+        bb[i] = g.bcentr .* tc["B"]
+    end
+    eq.vacuum_toroidal_field.b0 = bb
+    eq.vacuum_toroidal_field.r0 = gs[1].rcentr
+
+    for g in gs
+        geqdsk2imas!(
+            g, eq;
+            geqdsk_cocos=geqdsk_cocos, dd_cocos=dd_cocos, cocos_clockwise_phi=cocos_clockwise_phi,
+            add_derived=add_derived,
+        )
+    end
+end
+
+"""
+    geqdsk2imas!(
+        g::GEQDSKFile,
+        dd::IMASdd.dd;
+        geqdsk_cocos::Int=0,
+        dd_cocos::Int=11,
+        cocos_clockwise_phi::Bool=false,
+        add_derived::Bool=false,
+    )
+
+Utility for writing equilibrium data from a GEQDSKFile to IMAS DD.
+
+ * g: GEQDSKFile instance
+ * dd: IMAS data dictionary
+ * geqdsk_cocos: coordinate convention indentifier of the incoming GEQDSK set. 0 to auto detect.
+ * dd_cocos: coordinate convention identifier of the DD. Should normally be left at 11.
+ * cocos_clockwise_phi: hint for deciding COCOS, because GEQDSK probably doesn't constrain this.
+ * add_derived: switch to use functions within EFIT.jl to extend equilibrium data
+"""
+function geqdsk2imas!(
+    g::GEQDSKFile,
+    dd::IMASdd.dd;
+    geqdsk_cocos::Int=0,
+    dd_cocos::Int=11,
+    cocos_clockwise_phi::Bool=false,
+    add_derived::Bool=false,
+)
+    if ismissing(dd.dataset_description.data_entry, :pulse)
+        dd.dataset_description.data_entry.pulse = parse(Int, split(split(g.file, ".")[end-1], "g")[2])
+    end
+    geqdsk2imas!(
+        g, dd.equilibrium;
+        wall=dd.wall, geqdsk_cocos=geqdsk_cocos, dd_cocos=dd_cocos, cocos_clockwise_phi=cocos_clockwise_phi,
+        add_derived=add_derived,
+    )
```

Comment:
Not critical, but did you know that when the variable name matches the keyword argument name you can skip the keyword argument? Like this:
```suggestion
    geqdsk2imas!(
        g, dd.equilibrium;
        dd.wall, geqdsk_cocos, dd_cocos, cocos_clockwise_phi, add_derived,
    )
```